stan/math/rev/functor/gradient.hpp

```diff
@@ -51,6 +55,68 @@ void gradient(const F& f, const Eigen::Matrix<double, Eigen::Dynamic, 1>& x,
   grad_fx = x_var.adj();
 }
 
+/**
+ * Calculate the value and the gradient of the specified function
+ * at the specified argument.
+ *
+ * <p>The functor must implement
+ *
+ * <code>
+ * var
+ * operator()(const
+ * Eigen::Matrix<var, Eigen::Dynamic, 1>&)
+ * </code>
+ *
+ * using only operations that are defined for
+ * <code>var</code>.  This latter constraint usually
+ * requires the functions to be defined in terms of the libraries
+ * defined in Stan or in terms of functions with appropriately
+ * general namespace imports that eventually depend on functions
+ * defined in Stan.
+ *
+ * The evaluated gradient is stored into the object whose data
+ * begins at <code>*first_grad_fx</code> and ends at
+ * <code>*last_grad_fx</code>.  The caller is responsible for
+ * ensuring the size of the object pointed to by
+ * <code>first_grad_fx</code> matches the size of the argument
+ * <code>x</code>.
+ *
+ * <p>Time and memory usage is on the order of the size of the
+ * fully unfolded expression for the function applied to the
+ * argument, independently of dimension.
+ *
+ * @tparam F Type of function
+ * @param[in] f Function
+ * @param[in] x Argument to function
+ * @param[out] fx Function applied to argument
+ * @param[out] grad_fx Gradient of function at argument
+ * @throw std::invalid_argument if the iterator isn't the right size
+ * to hold the gradients
+ */
+template <typename F, typename InputIt>
+void gradient(
+    const F& f,
+    const Eigen::Ref<const Eigen::Matrix<double, Eigen::Dynamic, 1>>& x,
+    double& fx, InputIt first_grad_fx, InputIt last_grad_fx) {
```

Comment:
As long as `x` is only being used for 

```cpp
 Eigen::Matrix<var, Eigen::Dynamic, 1> x_var(x);
```

Then I think the below should work

```suggestion
template <typename F, typename EigVec, typename InputIt,
  require_eigen_vector_vt<std::is_arithmetic, EigVec>* = nullptr>
void gradient(const F& f, const EigVec& x, double& fx, InputIt first_grad_fx, InputIt last_grad_fx) {
```

Also you need docs for `InputIt`. I think it's fine to take the docs from cppreference like 

```
@tparam InputIt must meet the requirements of [LegacyInputIterator](https://en.cppreference.com/w/cpp/named_req/InputIterator).
```

Reply:
This is great.  Docs for sure.  Good catch.  You mind if I ask a few follow up questions first, just to make sure I understand all the things?

1. Understanding. Written this way the compiler will write a function that takes necessarily an Eigen vector in place of EigVec.  So if I pass in a Matrix it will write out a function for an Eigen Matrix.  And, importantly for my use case, if I pass in an Eigen Map, it will write a function for an Eigen Map and thus will avoid any copies (implicit or otherwise), right?
2. Limitations.  You say, "As long as x is only being used for `Eigen::Matrix<var, Eigen::Dynamic, 1> x_var(x);` then this should work."  If `const EigVec& x` is an Eigen vector, then we should be able to do anything with `x` we can do with an Eigen vector, right?  What other limitations do you have in mind?
3. Tests.  Presumably `require_eigen_vector_vt` is tested somewhere, so we do not need to test every new use of it, correct?

Reply:
No problem!


> Written this way the compiler will write a function that takes necessarily an Eigen vector in place of EigVec. So if I pass in a Matrix it will write out a function for an Eigen Matrix. And, importantly for my use case, if I pass in an Eigen Map, it will write a function for an Eigen Map and thus will avoid any copies (implicit or otherwise), right?

Yep! Eigen::Ref<> does something very similar, though we don't use it much in the math library (for a long time the docs and how it chose to make copies was pretty confusing to us). So I think the general template with a `require` restriction template to limit it to Eigen types with 1 compile time row or column would be good here.

>    Limitations. You say, "As long as x is only being used for Eigen::Matrix<var, Eigen::Dynamic, 1> x_var(x); then this should work." If const EigVec& x is an Eigen vector, then we should be able to do anything with x we can do with an Eigen vector, right? What other limitations do you have in mind?

So with `Eigen::Ref<>` directly it is true that your input can be treated as an Eigen::Vector with no worries, but if a function uses a general template for an Eigen type then we can end up passing Eigen expressions into functions. For example, doing the same coefficient access twice for an eigen expression will compute the expression twice. For functions that can take in expressions you need to use Stan's `to_ref()` function to do the equivalent of having `Eigen::Ref<>` in the signature.

But in your case we are just reading from the vector so idt it's necessary to use `to_ref()` here. 

> Tests. Presumably require_eigen_vector_vt is tested somewhere, so we do not need to test every new use of it, correct?

Yes no need to test a new use of it. You can add tests to check the row vectors now work as well, or you can just restrict your require to instead be `require_eigen_column_vector_vt<std::is_arithmetic, EigVec>` which will only support Eigen vectors